packages/go_router/example/lib/stateful_nested_navigation.dart

```diff
@@ -0,0 +1,420 @@
+// each tab.
+
+void main() {
+  runApp(NestedTabNavigationExampleApp());
+}
+
+/// An example demonstrating how to use nested navigators
+class NestedTabNavigationExampleApp extends StatelessWidget {
+  /// Creates a NestedTabNavigationExampleApp
+  NestedTabNavigationExampleApp({Key? key}) : super(key: key);
+
+  final GoRouter _router = GoRouter(
+    initialLocation: '/a',
+    restorationScopeId: 'sadfasf',
+    routes: <RouteBase>[
+      StatefulShellRoute(
+        branches: <ShellRouteBranch>[
+          /// The route branch for the first tab of the bottom navigation bar.
+          ShellRouteBranch(
+            navigatorKey: _tabANavigatorKey,
+            rootRoute: GoRoute(
+              /// The screen to display as the root in the first tab of the bottom
+              /// navigation bar.
+              path: '/a',
+              builder: (BuildContext context, GoRouterState state) =>
+                  const RootScreen(label: 'A', detailsPath: '/a/details'),
+              routes: <RouteBase>[
+                /// The details screen to display stacked on navigator of the
+                /// first tab. This will cover screen A but not the application
+                /// shell (bottom navigation bar).
+                GoRoute(
+                  path: 'details',
+                  builder: (BuildContext context, GoRouterState state) =>
+                      const DetailsScreen(label: 'A'),
+                ),
+              ],
+            ),
+          ),
+
+          /// The route branch for the second tab of the bottom navigation bar.
+          ShellRouteBranch(
+            /// It's not necessary to provide a navigatorKey if it isn't also
+            /// needed elsewhere. If not provided, a default key will be used.
+            // navigatorKey: _tabBNavigatorKey,
+            rootRoute: GoRoute(
+              /// The screen to display as the root in the second tab of the bottom
+              /// navigation bar.
+              path: '/b',
+              builder: (BuildContext context, GoRouterState state) =>
+                  const RootScreen(
+                      label: 'B',
+                      detailsPath: '/b/details/1',
+                      secondDetailsPath: '/b/details/2'),
+              routes: <RouteBase>[
+                GoRoute(
+                  path: 'details/:param',
+                  builder: (BuildContext context, GoRouterState state) =>
+                      DetailsScreen(label: 'B', param: state.params['param']),
+                ),
+              ],
+            ),
+          ),
+
+          /// The route branch for the third tab of the bottom navigation bar.
+          ShellRouteBranch(
+            /// Since this route branch has a nested StatefulShellRoute as the
+            /// root route, we need to specify what the default location for the
+            /// branch is.
+            defaultLocation: '/c1',
+            rootRoute: StatefulShellRoute.rootRoutes(
```

Comment:
A few questions here...

- When would a developer use  the `StatefulShellRoute.rootRoutes` constructor instead of the default?
- Why does one take a `List<ShellRouteBranch>` and the other takes a list of routes? 
- Is it okay to use the default `StatefulShellRoute` constructor here? 
- Are we detecting these types of misconfigurations somehow and throw an error up-front (for example, an assert in the constructor?)

Reply:
* Since `StatefulShellRoute.rootRoute` makes it possible to provide the root routes of the navigation branches directly, instead of going via `ShellRouteBranch`, it leads to somewhat more readable and simpler code when declaring routes (mostly since you can "save" one level one nesting). But this is only a shorthand/convenience constructor (for likely one of the more common use cases), so if you think it only adds noise, it can be removed. 
* The constructor that takes `List<ShellRouteBranch>` is the default way of configuring a `StatefulShellRoute`, to have the most configuration options (like not being restricted to `GoRoute`s and being able to provide a `restorationScopeId` and in the future; `NavigatorObservers`). The constructor that accepts GoRoutes is only a shorthand/convenience constructor (I tried to emphasise this more in the docs), that has some limitations, i.e. it lacks the configuration options the other constructor provides.
* Yes, the default constructor could be used here. 
* Yes, there are assertions in place in `StatefulShellRoute`, `ShellRouteBranch` and also in `RouteConfiguration`.